lib/SPIRV/libSPIRV/SPIRVModule.cpp

```diff
@@ -1210,6 +1301,42 @@ void SPIRVModuleImpl::addEntryPoint(SPIRVExecutionModelKind ExecModel,
   addCapabilities(SPIRV::getCapability(ExecModel));
 }
 
+void SPIRVModuleImpl::addConditionalEntryPoint(
+    SPIRVId Condition, SPIRVExecutionModelKind ExecModel, SPIRVId EntryPoint,
+    const std::string &Name, const std::vector<SPIRVId> &Variables) {
+  assert(isValid(ExecModel) && "Invalid execution model");
+  assert(EntryPoint != SPIRVID_INVALID && "Invalid entry point");
+  auto *EP = add(new SPIRVConditionalEntryPointINTEL(
+      this, Condition, ExecModel, EntryPoint, Name, Variables));
+  ConditionalEntryPointVec.push_back(EP);
+  ConditionalEntryPointSet[ExecModel].insert(EntryPoint);
+}
+
+void SPIRVModuleImpl::specializeConditionalEntryPoints(SPIRVId Condition,
+                                                       bool ShouldKeep) {
+  std::vector<std::pair<int, SPIRVId>> ToRemove;
+  for (unsigned It = 0; It < ConditionalEntryPointVec.size(); ++It) {
+    if (ConditionalEntryPointVec[It]->getCondition() == Condition) {
+      const auto *const EP = ConditionalEntryPointVec[It];
+      ToRemove.push_back(std::make_pair(It, EP->getTargetId()));
+      if (ShouldKeep) {
+        // add the removed conditional entry point as a normal entry point
+        addEntryPoint(EP->getExecModel(), EP->getTargetId(), EP->getName(),
+                      EP->getVariables());
+      }
+    }
+  }
+
+  for (const auto &It : ToRemove) {
+    const auto IRemove = It.first;
+    const auto Id = It.second;
+    ConditionalEntryPointVec.erase(ConditionalEntryPointVec.begin() + IRemove);
+    for (auto &[ExecMode, EPSet] : ConditionalEntryPointSet) {
+      EPSet.erase(Id);
+    }
+  }
+}
```

Comment:
It seems like UB to me: we have already mutated `ConditionalEntryPointVec` - yet we are using original indices stored in `ToRemove` container. Should we first sort `ToRemove` in descending order?